### ISSUE TITLE
Support for neural vocoders

### ISSUE THREAD

**Issue by YellowLime77** [nnsvs/nnsvs#72]
I've posted this on the oatsu-gh/ENUNU repository too (https://github.com/oatsu-gh/ENUNU/issues/11).

Just as the title says, I'm wondering if the code can be modified to use other vocoders, like the neural WaveNet vocoder. If so, steps could be provided which could really help me out, as even with a dataset with good quality and quantity of files, the final result always seems robotic, which I believe is because of the vocoder. Therefore, neural network vocoders could maybe produce better results.

Thank you so much!

**Comment by taroushirani** [comment]
Hello, I wrote the code to integrate NSF[1] to nnsvs once[2] (this code is for the old nnsvs and won't work with current one).  The key points are as follows:
(1) NNSVS stores the wav data and acoustic features generated from them under <directory of recipe>/dump/org, and you can use them to train any neural vocoder you want to use.
(2) Hack nnsvs/bin/synthesis.py to divide in two processes, parameter generation and waveform generation.   You can use this parameter to generate waveform by neural vocoder.

1. https://nii-yamagishilab.github.io/samples-nsf/
2. https://github.com/r9y9/nnsvs/pull/10

In my experience, the quality generated by NSF was not improved so much, so I think it is needed to improve acoustic model first.

**Comment by YellowLime77** [comment]
Thanks for replying! I will try to implement the steps and report back if I can get it to work (I'm not that proficient with programming).

**Comment by r9y9** [comment]
I would like to note that I plan to revisit the neural vocoder support after huge refactoring #81. I will be busy at work so It would take some time though. Technically, I think it's not that difficult.

Rough plans:
- Refactoring and pitch/duration improvements #81 
- Deploy and demo improvements #83 #78. I plan to implement a model packaging mechanism similar to https://github.com/r9y9/ttslearn.
- Revisit #10

**Comment by r9y9** [comment]
Let me rename the issue title.

**Comment by r9y9** [comment]
Code is ready at the dev branch: https://github.com/r9y9/nnsvs/tree/dev


Here's a pre-trained model using the [mixed/voc-multidb-latest](https://github.com/r9y9/nnsvs/tree/dev/recipes/mixed/voc-multidb-latest) recipe: https://www.dropbox.com/sh/52u8u5z41e4sdnz/AABTeOwtgyRiPqrAa5sjXFSHa?dl=0 

I trained the model by:

```
./run.sh  --stage 4 --stop-stage 4 --vocoder-model hn-sinc-nsf_sr48k_pwgD
```

and packaged the model by:

```
./run.sh  --stage 99 --stop-stage 99 --vocoder-model hn-sinc-nsf_sr48k_pwgD
```

NOTE: new updated recipes have extra packaging stage 99. Example: https://github.com/r9y9/nnsvs/blob/e6cc845102e97c6b46b9c06ce88d5578a9d93374/recipes/kiritan_singing/svs-world-resf0/run.sh#L153-L208. Packed models can be easily used by `nnsvs.svs.SPSVS` class (https://r9y9.github.io/nnsvs/svs.html).

You can use the pre-trained model (1) as a neural vocoder for your SVS (2) for fine-tuning your new vocoder. I will write a brief guide below.

## Using a pre-trained vocoder for your SVS system

Example: https://github.com/r9y9/nnsvs/tree/dev/recipes/nit-song070/svs-world-resf0

Steps:

1. Train timelag/duration/acoustic models as usual.
2. Run packaging process (stage 99) by explicitly specifying a pretrained vocoder path (`--vocoder-eval-checkpoint`). Check the following real command from my zsh history as an example:

```
CUDA_VISIBLE_DEVICES=1 ./run.sh  --stage 99 --stop-stage 99 --pretrained-expdir $PWD/../../kiritan_singing/svs-world-resf0/exp/kiritan_dim331-sv56 --tag dim331_sv56_pretrain_new_v2_allpre --acoustic_model acoustic_resf0convlstm --vocoder-eval-checkpoint $PWD/../../mixed/voc-multidb-latest/packed_models/mixed_hn-sinc-nsf_sr48k_pwgD/vocoder_model.pth
```

The packed model supports both WORLD and neural vocoder for waveform generation. See https://github.com/r9y9/nnsvs/blob/e6cc845102e97c6b46b9c06ce88d5578a9d93374/nnsvs/svs.py#L194

## Fine-tuning with a pre-trained vocoder

Example: https://github.com/r9y9/nnsvs/blob/dev/recipes/kiritan_singing/svs-world-resf0

Please specify ``--pretrained_vocoder_checkpoint`` and run the vocoder training step. After training, follow the same packaging process as mentioned above.

**Comment by r9y9** [comment]
Fixed by #139 

Docs: https://nnsvs.github.io/train_vocoders.html
